File: control-plane/consensus/consensus-feedback-loop.ts
#!/usr/bin/env node

/**
 * Consensus Feedback Loop - TypeScript/Node.js Implementation
 * Event-driven consensus-based agent orchestration for infrastructure optimization.
 * Leverages Node.js event loop and JavaScript async/await for non-blocking consensus operations.
 */

import { EventEmitter } from 'events';
import * as process from 'process';

// Data structures for consensus
interface ConsensusProposal {
  id: string;
  agentType: string;
  description: string;
  priority: number;
  timestamp: Date;
  votesRequired: number;
  approved: boolean;
}

interface ConsensusState {
  quorumSize: number;
  protocol: string;
  activeProposals: Map<string, ConsensusProposal>;
  lastUpdate?: Date;
}

// Agent interface
interface ConsensusAgent {
  localOptimization(state: ConsensusState): Promise<ConsensusProposal | null>;
  voteOnProposal(proposal: ConsensusProposal): Promise<boolean>;
  executeConsensus(proposal: ConsensusProposal, votes: boolean[]): Promise<void>;
}

// Cost optimizer agent implementation
class CostOptimizerAgent implements ConsensusAgent {
  constructor(private state: ConsensusState) {}

  async localOptimization(state: ConsensusState): Promise<ConsensusProposal | null> {
    // Simulate local optimization work
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Mock: Sometimes create a proposal (30% chance)
    if (Math.random() < 0.3) {
      return {
        id: `proposal-${Date.now()}`,
        agentType: 'cost-optimizer',
        description: 'Scale EKS node group from 3 to 5 nodes',
        priority: 2,
        timestamp: new Date(),
        votesRequired: this.state.quorumSize,
        approved: false
      };
    }

    return null;
  }

  async voteOnProposal(proposal: ConsensusProposal): Promise<boolean> {
    // Weighted voting based on agent type
    await new Promise(resolve => setTimeout(resolve, 100)); // Simulate voting deliberation

    if (proposal.agentType === 'cost-optimizer') {
      return true; // Cost optimizer always approves cost-related proposals
    }

    // For other proposals, approve with 80% confidence
    return Math.random() < 0.8;
  }

  async executeConsensus(proposal: ConsensusProposal, votes: boolean[]): Promise<void> {
    console.log(`Executing consensus proposal: ${proposal.description}`);

    // Simulate infrastructure change execution
    await new Promise(resolve => setTimeout(resolve, 500));

    console.log('Consensus change executed successfully');
  }
}

// Main consensus feedback loop
class ConsensusFeedbackLoop extends EventEmitter {
  private consensusState: ConsensusState;
  private agents: ConsensusAgent[];
  private feedbackInterval: number;
  private quorumSize: number;
  private running: boolean = true;

  constructor() {
    super();

    // Read environment variables
    this.consensusState = {
      quorumSize: parseInt(process.env.AGENT_QUORUM || '3'),
      protocol: process.env.CONSENSUS_PROTOCOL || 'raft',
      activeProposals: new Map()
    };

    this.feedbackInterval = parseInt((process.env.FEEDBACK_INTERVAL || '30s').replace('s', ''));
    this.quorumSize = this.consensusState.quorumSize;

    // Initialize agents
    this.agents = [
      new CostOptimizerAgent(this.consensusState)
      // Add more agent types as needed
    ];

    // Setup signal handling for graceful shutdown
    process.on('SIGTERM', () => {
      console.log('Received SIGTERM, shutting down gracefully...');
      this.running = false;
    });

    process.on('SIGINT', () => {
      console.log('Received SIGINT, shutting down gracefully...');
      this.running = false;
    });
  }

  async runFeedbackLoop(): Promise<void> {
    console.log('Starting TypeScript consensus feedback loop');

    while (this.running) {
      try {
        // Phase 1: Local optimization (30 seconds)
        await this.runLocalOptimizationPhase();

        // Phase 2: Consensus proposal submission
        await this.runConsensusProposalPhase();

        // Phase 3: Vote collection
        await this.runVoteCollectionPhase();

        // Phase 4: Consensus execution
        await this.runConsensusExecutionPhase();

        // Wait for next iteration
        await new Promise(resolve => setTimeout(resolve, (60 - this.feedbackInterval) * 1000));
      } catch (error) {
        console.error(`Error in feedback loop iteration: ${(error as Error).message}`);
        await new Promise(resolve => setTimeout(resolve, 10000));
      }
    }
  }

  private async runLocalOptimizationPhase(): Promise<void> {
    console.log('Starting local optimization phase');

    try {
      // Run local optimization with timeout
      const optimizationTasks = this.agents.map(agent =>
        this.withTimeout(agent.localOptimization(this.consensusState), this.feedbackInterval * 1000)
      );

      const optimizationResults = await Promise.allSettled(optimizationTasks);

      for (const result of optimizationResults) {
        if (result.status === 'fulfilled' && result.value) {
          this.consensusState.activeProposals.set(result.value.id, result.value);
          console.log(`Created consensus proposal: ${result.value.id}`);
        }
      }
    } catch (error) {
      console.error(`Local optimization failed: ${(error as Error).message}`);
    }
  }

  private async runConsensusProposalPhase(): Promise<void> {
    if (this.consensusState.activeProposals.size === 0) {
      return;
    }

    for (const [proposalId, proposal] of this.consensusState.activeProposals) {
      try {
        // Submit to consensus protocol (simulated)
        await this.submitToConsensus(proposal);
        console.log(`Submitted proposal ${proposalId} to consensus`);
      } catch (error) {
        console.error(`Failed to submit proposal ${proposalId}: ${(error as Error).message}`);
      }
    }
  }

  private async runVoteCollectionPhase(): Promise<void> {
    for (const [proposalId, proposal] of this.consensusState.activeProposals) {
      try {
        const votes = await this.collectVotes(proposal);

        if (votes.filter(v => v).length >= proposal.votesRequired) {
          proposal.approved = true;
          console.log(`Proposal ${proposalId} reached quorum`);
        } else if (votes.length > 0) {
          console.log(`Proposal ${proposalId} has ${votes.filter(v => v).length} votes`);
        }
      } catch (error) {
        console.error(`Failed to collect votes for proposal ${proposalId}: ${(error as Error).message}`);
      }
    }
  }

  private async runConsensusExecutionPhase(): Promise<void> {
    for (const [proposalId, proposal] of this.consensusState.activeProposals) {
      if (proposal.approved) {
        try {
          // Collect final votes for execution
          const finalVotes = await this.collectVotes(proposal);
          await this.agents[0].executeConsensus(proposal, finalVotes);

          this.consensusState.activeProposals.delete(proposalId);
          console.log(`Executed approved proposal: ${proposalId}`);
        } catch (error) {
          console.error(`Failed to execute proposal ${proposalId}: ${(error as Error).message}`);
        }
      }
    }
  }

  private async submitToConsensus(proposal: ConsensusProposal): Promise<void> {
    // Simulate consensus protocol submission
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  private async collectVotes(proposal: ConsensusProposal): Promise<boolean[]> {
    // Simulate vote collection from agents
    const votes: boolean[] = [];

    for (const agent of this.agents) {
      try {
        const vote = await agent.voteOnProposal(proposal);
        votes.push(vote);
      } catch (error) {
        console.error(`Agent voting failed: ${(error as Error).message}`);
        votes.push(false); // Default to false on error
      }
    }

    return votes;
  }

  private async withTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Operation timed out'));
      }, timeoutMs);

      promise
        .then(resolve)
        .catch(reject)
        .finally(() => clearTimeout(timeout));
    });
  }
}

// Main execution
async function main() {
  const loop = new ConsensusFeedbackLoop();

  try {
    await loop.runFeedbackLoop();
  } catch (error) {
    console.error(`Consensus feedback loop failed: ${(error as Error).message}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { ConsensusFeedbackLoop, ConsensusProposal, ConsensusState, ConsensusAgent };

File: core/operators/control-plane/consensus/consensus_feedback_loop.py
#!/usr/bin/env python3
"""
Consensus Feedback Loop - Python Implementation
Object-oriented consensus-based agent orchestration for infrastructure optimization.
Provides better error handling and structure than bash while remaining runtime-agnostic.
"""

import os
import time
import signal
import logging
import asyncio
from typing import Optional, Dict, Any
from dataclasses import dataclass
from datetime import datetime, timedelta

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

@dataclass
class ConsensusProposal:
    """Represents a consensus proposal for infrastructure changes"""
    id: str
    agent_type: str
    description: str
    priority: int
    timestamp: datetime
    votes_required: int = 3

@dataclass
class ConsensusState:
    """Current state of the consensus process"""
    quorum_size: int = 3
    protocol: str = "raft"
    active_proposals: Dict[str, ConsensusProposal] = None
    last_update: Optional[datetime] = None

    def __post_init__(self):
        if self.active_proposals is None:
            self.active_proposals = {}

class ConsensusFeedbackLoop:
    """Main consensus feedback loop implementation"""

    def __init__(self):
        self.consensus_enabled = os.getenv('CONSENSUS_ENABLED', 'false').lower() == 'true'
        self.feedback_interval = int(os.getenv('FEEDBACK_INTERVAL', '30').rstrip('s'))
        self.agent_quorum = int(os.getenv('AGENT_QUORUM', '3'))
        self.consensus_protocol = os.getenv('CONSENSUS_PROTOCOL', 'raft')

        self.consensus_state = ConsensusState(
            quorum_size=self.agent_quorum,
            protocol=self.consensus_protocol
        )

        # Setup signal handling for graceful shutdown
        signal.signal(signal.SIGTERM, self._signal_handler)
        signal.signal(signal.SIGINT, self._signal_handler)

        self.running = True

    def _signal_handler(self, signum, frame):
        """Handle shutdown signals gracefully"""
        logger.info(f"Received signal {signum}, shutting down gracefully...")
        self.running = False

    async def run_feedback_loop(self):
        """Main feedback loop - runs every minute with 30s optimization windows"""
        logger.info("Starting consensus feedback loop")

        while self.running:
            try:
                # Local optimization phase (30 seconds)
                await self._local_optimization_phase()

                # Consensus proposal phase
                await self._consensus_proposal_phase()

                # Vote collection phase
                await self._vote_collection_phase()

                # Change execution phase
                await self._change_execution_phase()

                # Wait for next iteration (remaining time in minute)
                await asyncio.sleep(max(0, 60 - self.feedback_interval))

            except Exception as e:
                logger.error(f"Error in feedback loop iteration: {e}")
                await asyncio.sleep(10)  # Brief pause before retry

    async def _local_optimization_phase(self):
        """Phase 1: Local agent optimization with 30s timeout"""
        logger.debug("Starting local optimization phase")

        try:
            # Create timeout task for local optimization
            optimization_task = asyncio.create_task(self._run_local_optimization())
            result = await asyncio.wait_for(optimization_task, timeout=self.feedback_interval)

            if result and result.get('proposal_required', False):
                await self._create_consensus_proposal(result)

        except asyncio.TimeoutError:
            logger.warning("Local optimization timed out")
        except Exception as e:
            logger.error(f"Local optimization failed: {e}")

    async def _run_local_optimization(self) -> Optional[Dict[str, Any]]:
        """Run local agent optimization logic"""
        # This would interface with the actual agent optimization code
        # For now, return mock optimization results
        await asyncio.sleep(1)  # Simulate optimization work

        # Mock: Sometimes create a proposal
        import random
        if random.random() < 0.3:  # 30% chance of proposal
            return {
                'proposal_required': True,
                'agent_type': 'cost-optimizer',
                'description': 'Scale EKS node group from 3 to 5 nodes',
                'priority': 2
            }

        return {'proposal_required': False}

    async def _create_consensus_proposal(self, optimization_result: Dict[str, Any]):
        """Create a consensus proposal from optimization results"""
        proposal = ConsensusProposal(
            id=f"proposal-{int(time.time())}",
            agent_type=optimization_result['agent_type'],
            description=optimization_result['description'],
            priority=optimization_result['priority'],
            timestamp=datetime.now(),
            votes_required=self.agent_quorum
        )

        self.consensus_state.active_proposals[proposal.id] = proposal
        logger.info(f"Created consensus proposal: {proposal.id}")

    async def _consensus_proposal_phase(self):
        """Phase 2: Submit proposals to consensus protocol"""
        if not self.consensus_state.active_proposals:
            return

        for proposal_id, proposal in list(self.consensus_state.active_proposals.items()):
            try:
                # Submit to consensus protocol (Raft-based)
                await self._submit_to_consensus(proposal)
                logger.info(f"Submitted proposal {proposal_id} to consensus")

            except Exception as e:
                logger.error(f"Failed to submit proposal {proposal_id}: {e}")

    async def _submit_to_consensus(self, proposal: ConsensusProposal):
        """Submit proposal to Raft consensus protocol"""
        # This would interface with the Raft consensus implementation
        await asyncio.sleep(0.1)  # Simulate consensus submission

    async def _vote_collection_phase(self):
        """Phase 3: Collect votes from agent quorum"""
        for proposal_id, proposal in list(self.consensus_state.active_proposals.items()):
            try:
                votes = await self._collect_votes(proposal)

                if len(votes) >= proposal.votes_required:
                    # Quorum reached - mark for execution
                    proposal.approved = True
                    logger.info(f"Proposal {proposal_id} reached quorum")
                elif len(votes) > 0:
                    # Partial votes - keep collecting
                    logger.debug(f"Proposal {proposal_id} has {len(votes)} votes")
                else:
                    # No votes yet - continue
                    pass

            except Exception as e:
                logger.error(f"Failed to collect votes for proposal {proposal_id}: {e}")

    async def _collect_votes(self, proposal: ConsensusProposal) -> list:
        """Collect votes from agent quorum"""
        # This would query the consensus protocol for votes
        await asyncio.sleep(0.1)  # Simulate vote collection

        # Mock: Return some votes
        import random
        vote_count = random.randint(0, self.agent_quorum)
        return [f"agent-{i}" for i in range(vote_count)]

    async def _change_execution_phase(self):
        """Phase 4: Execute consensus-approved changes"""
        for proposal_id, proposal in list(self.consensus_state.active_proposals.items()):
            if getattr(proposal, 'approved', False):
                try:
                    await self._execute_changes(proposal)
                    del self.consensus_state.active_proposals[proposal_id]
                    logger.info(f"Executed approved proposal: {proposal_id}")

                except Exception as e:
                    logger.error(f"Failed to execute proposal {proposal_id}: {e}")

    async def _execute_changes(self, proposal: ConsensusProposal):
        """Execute the approved infrastructure changes"""
        # This would interface with the actual infrastructure change execution
        logger.info(f"Executing infrastructure change: {proposal.description}")
        await asyncio.sleep(0.5)  # Simulate change execution

def main():
    """Main entry point"""
    loop = ConsensusFeedbackLoop()

    try:
        asyncio.run(loop.run_feedback_loop())
    except KeyboardInterrupt:
        logger.info("Consensus feedback loop stopped by user")
    except Exception as e:
        logger.error(f"Consensus feedback loop failed: {e}")
        raise

if __name__ == "__main__":
    main()
